src/replication.c

```diff
@@ -687,17 +730,61 @@ replica_receiver(void *arg) {
 	return NULL;
 }
 
+void cleanup_replica(replica_t *replica) {
+	MTX_LOCK(&replica->r_mtx);
+	if(replica->mgmt_fd != -1)  {
+		close(replica->mgmt_fd);
+		replica->mgmt_fd = -1;
+	}
+	if(replica->iofd != -1)  {
+		close(replica->iofd);
+		replica->iofd = -1;
+	}
+	if(replica->sender_epfd != -1) {
+		close(replica->sender_epfd);
+		replica->sender_epfd = -1;
+	}
+	__sync_fetch_and_sub(&replica->refcount, 1);
+	if(replica->refcount != 0) {
+		MTX_UNLOCK(&replica->r_mtx);
+		return;
+	}
+	if(replica->io_resp_hdr) {
+		free(replica->io_resp_hdr);
+	}
+	if(replica->ip) {
+		free(replica->ip);
+	}
+	free(replica->mgmt_ack);
+	free(replica->mgmt_ack_data);
+	MTX_UNLOCK(&replica->r_mtx);
+	release_replica_id(replica->spec, replica->id);
+	free(replica);
+}
+
 /* creates replica entry and adds to spec's rwaitq list after creating mgmt connection */
 replica_t *
 create_replica_entry(spec_t *spec, int mgmt_fd)
 {
 	replica_t *replica = (replica_t *)malloc(sizeof(replica_t));
+	replica->refcount = 0;
+	__sync_fetch_and_add(&replica->refcount, 1);
 	replica->mgmt_fd = mgmt_fd;
+	replica->iofd = -1;
+	replica->io_resp_hdr = NULL;
+	replica->ip = NULL;
 	replica->mgmt_io_state = READ_IO_RESP_HDR;
 	replica->mgmt_io_read = 0;
 	replica->mgmt_ack = (zvol_io_hdr_t *)malloc(sizeof(zvol_io_hdr_t));
 	replica->mgmt_ack_data = (mgmt_ack_data_t *)malloc(sizeof(mgmt_ack_data_t));
 	MTX_LOCK(&spec->rq_mtx);
+	replica->id = allocate_replica_id(spec);
```

Comment:
with latest changes, replica 'id' may not be required. While merging, this can be removed completely

Reply:
While merging, this can be removed completely if its not in use anymore